mesonbuild/options.py

```diff
@@ -483,7 +483,31 @@ def is_auto(self) -> bool:
         return self.value == 'auto'
 
 
-@dataclasses.dataclass(init=False)
+_U = T.TypeVar('_U', bound=UserOption)
+
+
+def choices_are_different(a: _U, b: _U) -> bool:
+    """Are the choices between two options the same?
+
+    :param a: A UserOption[T]
+    :param b: A second UserOption[T]
+    :return: True if the choices have changed, otherwise False
+    """
+    if isinstance(a, EnumeratedUserOption):
+        # If this isn't hte case the above failed somehow
+        assert isinstance(b, EnumeratedUserOption), 'for mypy'
```

Comment:
Is mypy really this dumb? Isn't `isinstance()` recognized as a type narrowing check? I don't know what else is then...

Reply:
It is a type narrowing check, but only for `a`, `b` is still left as `_U`, and even when I tried adding things like:
```python
assert type(a) is type(b)
```
mypy wasn't smart enough to figure out that if `a` is `EnumeratedUserOption` that `b` would also be `EnumeratedUserOption`.

Reply:
Oh, sorry, I misunderstood how type vars work when they reference unions. I though that if the same type vars appears in a function signature it is forced to resolve to the same concrete type. Evidently it does not work like this.

Reply:
Maybe it should? It would in rust and C++ wouldn't it? Maybe I should open a typing issue about this.